contracts/ReputationMiningCycle.sol

```diff
@@ -50,77 +79,110 @@ contract ReputationMiningCycle {
-    reputationMiningWindowOpenTimestamp = now;
-  }
-
-  function respondToChallenge(uint256 round, uint256 idx) public {
-    // TODO: Check challenge response is valid, relating to current challenge
-    // TODO: Check challenge response relates to this hash
-
-    // Assuming that the challenge response is correct...
-    uint256 opponentIdx = (idx % 2 == 1 ? idx-1 : idx + 1);
-
-    disputeRounds[round][idx].lastResponseTimestamp = now;
-    disputeRounds[round][idx].challengeStepCompleted += 1;
-    // If our opponent responded to this challenge before we did, we should
-    // reset their 'last response' time to now, as they aren't able to respond
-    // to the next challenge before they know what it is!
-    if (disputeRounds[round][idx].challengeStepCompleted == disputeRounds[round][opponentIdx].challengeStepCompleted) {
-      disputeRounds[round][opponentIdx].lastResponseTimestamp = now;
-    }
+  modifier challengeOpen(uint256 round, uint256 idx) {
+    // TODO: More checks that this is an appropriate time to respondToChallenge
+    require(disputeRounds[round][idx].lowerBound == disputeRounds[round][idx].upperBound);
+    _;
   }
 
-  function submitNewHash(bytes32 newHash, uint256 nNodes, uint256 entry) public {
-    //Check the ticket is an eligible one for them to claim
+  modifier entryQualifies(bytes32 newHash, uint256 nNodes, uint256 entry) {
+    // Check the ticket is an eligible one for them to claim
+    // TODO: Require minimum stake, that is (much) more than the cost required to defend the valid submission.
     require(entry <= IColonyNetwork(colonyNetworkAddress).getStakedBalance(msg.sender) / 10**15);
     require(entry > 0);
     if (reputationHashSubmissions[msg.sender].hash != 0x0) {           // If this user has submitted before during this round...
       require(newHash == reputationHashSubmissions[msg.sender].hash);  // ...require that they are submitting the same hash ...
       require(nNodes == reputationHashSubmissions[msg.sender].nNodes); // ...require that they are submitting the same number of nodes for that hash ...
       require (submittedEntries[newHash][msg.sender][entry] == false); // ... but not this exact entry
     }
-    // TODO: Require minimum stake, that is (much) more than the cost required to defend the valid submission.
+    _;
+  }
+
+  // A submission will only be accepted from a miner if `keccak256(address, N, hash) < target`
+  // At the beginning of the submission window, the target is set to 0 and slowly increases to 2^256 - 1 after an hour
+  modifier withinTarget(bytes32 newHash, uint256 entry) {
     // Check the ticket is a winning one.
     // TODO Figure out how to uncomment the next line, but not break tests sporadically.
     // require((now-reputationMiningWindowOpenTimestamp) <= 3600);
     // x = floor(uint((2**256 - 1) / 3600)
-    if (now-reputationMiningWindowOpenTimestamp <= 3600) {
+    if (now - reputationMiningWindowOpenTimestamp <= 3600) {
```

Comment:
Happy if we removed the todo above and just made it clear where `x` comes from in a comment. There's no need to perform the calculation every time

Reply:
The TODO is related to the `require` enforcing the finite window on reputation submissions. The line `// x = floor(uint((2**256 - 1) / 3600) ` is intended to be the comment explaining where `x` comes from.

Reply:
Yeah. There's a TODO in the mining client that should be able to reduce it to a vanishingly small chance (like I did in the `should allow someone to submit a new reputation hash if they are eligible inside the window` test). Then we can move some/all `forwardTime`s to `3540` or a similar number. Just hasn't been done yet, and given the number of tests that will require changing I'd rather do it in a separate PR.